### PR TITLE
rafthttp: add stream metrics

### PR DESCRIPTION
**Pull request by yichengq**
http endpoint: `/v2/stats/transport`
format:
```
unihorn@CoreOS ~/g/s/g/a/kafka> curl http://127.0.0.1:4002/v2/stats/transport | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   110    0   110    0     0  89576      0 --:--:-- --:--:-- --:--:--  107k
{
    "924e2e83e93f2560": {
        "stream": {
            "bytes_recieved": 385,
            "entries_received": 5,
            "last_index_received": 18307
        }
    }
}
unihorn@CoreOS ~/g/s/g/a/kafka> curl http://127.0.0.1:4003/v2/stats/transport | python -m json.tool
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   200    0   200    0     0   131k      0 --:--:-- --:--:-- --:--:--  195k
{
    "6e3bd23ae5f1eae0": {
        "stream": {
            "bytes_sent": 1078,
            "entries_sent": 14,
            "last_index_sent": 18316
        }
    },
    "a8266ecf031671f3": {
        "stream": {
            "bytes_sent": 1078,
            "entries_sent": 14,
            "last_index_sent": 18316
        }
    }
}
```

### CODE REVIEW

**Comment by yichengq** [comment]
for #1766

**Comment by yichengq** [comment]
@xiang90 refreshed.

**Comment by yichengq** [comment]
@xiang90 updated.

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+// Gauge returns instantaneous value that is expected to fluctuate over time.
+type Gauge interface {
+	Set(value int64)
+	String() string
+}
+
+// Group aggregates counters, gauges and sub-groups.
+type Group struct {
+	// clear operation needs exclusive access to map
+	mu sync.RWMutex
```
Comment (xiang90) [inline pkg/metrics/metrics.go]: mu sync.RWMutex -> sync.RWMutex

remove necessary comment

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+			fmt.Fprintf(&b, ", ")
+		}
+		fmt.Fprintf(&b, "%q: %v", kv.Key, v)
+		first = false
+	})
+	fmt.Fprintf(&b, "}")
+	return b.String()
+}
+
+type counter struct {
```
Comment (xiang90) [inline pkg/metrics/metrics.go]: one line

File: rafthttp/streamer.go [hunk 3]
```diff
@@ -278,9 +291,13 @@ func (s *streamReader) isStopped() bool {
 func (s *streamReader) handle(r io.Reader) {
 	defer func() {
 		close(s.done)
+		s.mg.Clear()
 		log.Printf("rafthttp: client streaming to %s at term %d has been stopped", s.to, s.term)
 	}()
 
+	entriesRecv := s.mg.Counter("entries_received")
```
Comment (xiang90) [inline rafthttp/streamer.go]: move the metrics bookkeeping to reader/writer

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (yichengq) [inline pkg/metrics/metrics.go]: i don't want to export Lock and Unlock functions, so i set it as private member.
comment removed.

File: pkg/metrics/metrics.go (hunk 2 above)
Comment (yichengq) [inline pkg/metrics/metrics.go]: done

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (xiang90) [inline pkg/metrics/metrics.go]: ok

File: rafthttp/streamer.go (hunk 3 above)
Comment (yichengq) [inline rafthttp/streamer.go]: done.

File: rafthttp/peer.go [hunk 4]
```diff
@@ -181,6 +184,7 @@ func (p *peer) Stop() {
 	p.Lock()
 	defer p.Unlock()
 	p.stream.stop()
+	p.mg.Clear()
```
Comment (xiang90) [inline rafthttp/peer.go]: can we do the clear in stream.Stop?

File: rafthttp/streamer.go [hunk 5]
```diff
@@ -153,23 +155,24 @@ type WriteFlusher interface {
...
 // The caller should call stop when finished, to shut it down.
 func newStreamWriter(w WriteFlusher, to types.ID, term uint64) *streamWriter {
 	s := &streamWriter{
+		w:    w,
 		to:   to,
 		term: term,
 		q:    make(chan []raftpb.Entry, streamBufSize),
 		done: make(chan struct{}),
 	}
-	go s.handle(w)
```
Comment (xiang90) [inline rafthttp/streamer.go]: why do we move this out of new function?

File: pkg/metrics/metrics.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
+/*
```
Comment (xiang90) [inline pkg/metrics/metrics.go]: we need test for this pkg

File: rafthttp/streamer.go (hunk 5 above)
Comment (yichengq) [inline rafthttp/streamer.go]: streamWriter needs to be set metricsGroup before handle goroutine starts(or there will be race), while streamWriter is inited in http handler, who doesn't know about the metricsGroup to use.

File: rafthttp/peer.go (hunk 4 above)
Comment (yichengq) [inline rafthttp/peer.go]: yes.

File: pkg/metrics/metrics.go (hunk 6 above)
Comment (yichengq) [inline pkg/metrics/metrics.go]: will add

**Comment by yichengq** [comment]
metric test added

File: rafthttp/entry_reader.go [hunk 7]
```diff
@@ -20,11 +20,24 @@ import (
...
 	"io"
 
+	"github.com/coreos/etcd/pkg/metrics"
 	"github.com/coreos/etcd/raft/raftpb"
 )
 
 type entryReader struct {
-	r io.Reader
+	r             io.Reader
+	entriesRecv   metrics.Counter
```
Comment (xiang90) [inline rafthttp/entry_reader.go]: totalEnts
totalBytes
lastIndex

File: rafthttp/entry_writer.go [hunk 8]
```diff
@@ -20,11 +20,24 @@ import (
...
 	"io"
 
+	"github.com/coreos/etcd/pkg/metrics"
 	"github.com/coreos/etcd/raft/raftpb"
 )
 
 type entryWriter struct {
-	w io.Writer
+	w             io.Writer
+	entriesSent   metrics.Counter
```
Comment (xiang90) [inline rafthttp/entry_writer.go]: totalEnts
totalBytes
lastIndex

File: rafthttp/entry_reader.go (hunk 7 above)
Comment (xiang90) [inline rafthttp/entry_reader.go]: well. probably we can also remove total. make it consistent with line37 but without received

**Comment by xiang90** [comment]
@yichengq How about this:
having a global group called etcd.
each etcd member has a subgroup named as its id. 
then we do not need to pass the group thing around.

**Comment by xiang90** [comment]
also we need to add /pkg/metrics to test.

File: pkg/metrics/metrics.go [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+	g.mu.RLock()
+	defer g.mu.RUnlock()
+	gg := &gauge{
+		i: new(expvar.Int),
+	}
+	g.m.Set(key, gg)
+	return gg
+}
+
+func (g *Group) Group(key string) *Group {
```
Comment (xiang90) [inline pkg/metrics/metrics.go]: NewGroup

**Comment by yichengq** [comment]
i will rewrite it based on discuss

**Pull request closed without merging**